sympy/core/expr.py

```diff
@@ -689,44 +689,28 @@ def equals(self, other, failing_expression=False):
         return None
 
     def _eval_is_positive(self):
-        from sympy.polys.numberfields import minimal_polynomial
-        from sympy.polys.polyerrors import NotAlgebraic
-        if self.is_number:
-            if self.is_real is False:
-                return False
-            try:
-                # check to see that we can get a value
-                n2 = self._eval_evalf(2)
-                if n2 is None:
-                    raise AttributeError
-                if n2._prec == 1:  # no significance
-                    raise AttributeError
-                if n2 == S.NaN:
-                    raise AttributeError
-            except (AttributeError, ValueError):
-                return None
-            n, i = self.evalf(2).as_real_imag()
-            if not i.is_Number or not n.is_Number:
-                return False
-            if n._prec != 1 and i._prec != 1:
-                return bool(not i and n > 0)
-            elif n._prec == 1 and (not i or i._prec == 1) and \
-                    self.is_algebraic and not self.has(Function):
-                try:
-                    if minimal_polynomial(self).is_Symbol:
-                        return False
-                except (NotAlgebraic, NotImplementedError):
-                    pass
+        positive = lambda x: x > 0
+        return self._eval_is_negative_or_positive(positive)
 
     def _eval_is_negative(self):
+        negative = lambda x: x < 0
+        return self._eval_is_negative_or_positive(negative)
+
+    def _eval_is_negative_or_positive(self, compare_sign):
         from sympy.polys.numberfields import minimal_polynomial
         from sympy.polys.polyerrors import NotAlgebraic
         if self.is_number:
             if self.is_real is False:
                 return False
+            # It is not infinity, so it can be calculated
             try:
-                # check to see that we can get a value
-                n2 = self._eval_evalf(2)
+                prec = 2
```

Comment:
If you search for 'big step' or something like that in the evalf file, you may find a better way to increase precision; you may also find a way to look up the current max_prec rather than assuming it is 256.

Reply:
> a better way to increase precision
```python
dp = 1
while 1:
    calc at prec
    if good enough:
        break
    prec += max(1, min(dp, 30, DEFAULT_MAXPREC - prec))
    if prec > DEFAULT_MAXPREC:
        raise PrecisionExhausted
    dp *= 2
```

Reply:
It would be handy to have a generator somewhere that does this so it's just:
```python
for fv in increasing_precision(expr):
    if good enough:
       break/return etc.
else:
    raise ValueError
```